doc/developer/design/20220404_metadata_storage.md

```diff
@@ -0,0 +1,304 @@
+  the metadata database therefore need to exhibit p95 commit latencies on the
+  order of single-digit milliseconds.
+
+* **Easy scalability.** The metadata database must scale to support tens of
+  thousands of customer environments in a cloud provider region without undue
+  operational burden.
+
+* **Compatibility with single binary and bring-your-own-cloud deployments.** The
+  metadata database must provide an API that is easily replicable with SQLite in
+  single binary deployments. It must also be easy to operate in
+  bring-your-own-cloud deployments.
+
+* **Environment schema isolation.** Version upgrades and rollbacks in one
+  environment must not be dependent on version upgrades and rollbacks in other
+  environments.
+
+* **Security sandboxing.** The database must enforce that an environment can
+  only access the data that belongs to that environment.
+
+### Desirables
+
+The following would be nice to have, but are not strict requirements.
+
+* **Consistency across clouds.** The ideal database would have an equivalent
+  offering across all of the major cloud providers.
+
+* **Ability to run in test environments.** The ideal database would be easy
+  to run in the Antithesis and Jepsen test environments to permit fault
+  injection.
+
+* **Environment performance isolation.** The ideal database would provide
+  per-environment performance isolation, so that an active or malicious
+  environment would not impact the performance of other environments.
+
+* **Migration plan to STORAGE collections.** In the far future, Materialize may
+  be able to use its own [STORAGE](../platform/architecture-db.md#storage) layer
+  to store its metadata. The ideal database would not inhibit this future.
+
+### Nonrequirements
+
+The following are explicitly not requirements of the metadata storage layer.
+
+* **Resilience to region outages.** The metadata storage layer does not need to
+  tolerate the simultaneous failure of multiple availability zones within a
+  cloud provider region.
+
+* **Distribution across multiple regions.** The metadata storage layer does not
+  need to support queries across multiple regions. We are comfortable operating
+  a separate metadata storage layer per region.
+
+* **Cross-environment reporting queries.** The metadata storage layer does not
+  need to support queries across multiple customer environments for
+  introspection or reporting.
+
+## Discussion
+
+### Current state of affairs
+
+Materialize Platform currently stores metadata in SQLite databases on an EBS
+volume. The hard requirements immediately rule this situation out as a viable
+long term option ([details][sqlite-problems]).
+
+### Abstraction
+
+`materialized` is moving towards storing all its metadata via the [`Stash`] API,
+which abstracts over any particular database's API. This API is intentionally
+kept simple enough that `Stash` can, in principle, be backed by any of the
+following:
```

Comment:
I'm inferring from this that we already recognize a couple of different sets of data that we want to treat differently:
 - catalog changes: strongly consistent, very fast, small data
 - audit refs to the above: eventually consistent, fast, medium to large data (probably want to sync it elsewhere, e.g. dumps to S3)
 
Although I'm also seeing this as a desire to keep things colocated for now for simplicity (which I'm sympathetic to). So I'm comfortable saying we want all data to be treated as if it was catalog change data, and we will optimize for other types down the road.

Reply:
Hmm, not sure—what I'm trying to say is that there is a most audit data _is_ catalog data, and so 100% needs to start its life in the catalog, where it can be inserted in the same transaction as the change to the catalog data. I don't see how you could do anything else if you want the audit data to be 100% correct (which we do for at least changes to cluster configuration, to power billing).